### DISCUSSION TITLE
Are there any examples of REST handlers with mutations?

### DISCUSSION THREAD

**Discussion by andyjessop** [mswjs/msw#1604] [Q&A] [answered]
All the examples I see are usually GET requests or simple login/logout with static fixtures as responses. But testing mutations requires more complex mocks. Do you have any examples of complex apps that use MSW as their primary dev backend API? If not, what strategies should I use for mocking mutations?

**Comment by kettanaito** [comment, marked as answer]
Hi, @andyjessop. Yes, you can use MSW to mock mutations as well. I don't have any examples to share from the top of my head but I know of many folks who're using MSW exactly for what you described. Ask around on Twitter and I'm sure people will share their setups and knowledge. 

When it comes to more complex mocks, I think the most essential part is the persistence of data. You can use whatever you want for persistence, e.g. global variables, `localStorage`, `IndexedDb`, and even external storages—as long as their application makes sense for what you're building. We also have some libraries in the ecosystem to scaffold more complex data-driven mocks—take a look at [mswjs/data](https://github.com/mswjs/data).

**Reply by kettanaito** [reply]
Oh, I do know one example: check out https://github.com/kentcdodds/kentcdodds.com. Kent builds and tests his entire website with MSW. I'm certain you can find some more advanced patterns there.

**Reply by andyjessop** [reply]
That's fantastic, cheers.
